python.d/linux_power_supply.chart.py

```diff
@@ -0,0 +1,170 @@
+    attr_max_design = '{0}_{1}_design'.format(name, maxname)
+    attr_max = '{0}_{1}'.format(name, maxname)
+    attr_now = '{0}_now'.format(name)
+    attr_min = '{0}_{1}'.format(name, minname)
+    attr_min_design = '{0}_{1}_design'.format(name, minname)
+    if get_sysfs_value(os.path.join(syspath, attr_now)) is not None:
+        lines.append([attr_now, attr_now, 'absolute', 1, PRECISION])
+        attrlist.append(attr_now)
+    else:
+        return None, None
+    if get_sysfs_value(os.path.join(syspath, attr_max)) is not None:
+        lines.insert(0, [attr_max, attr_max, 'absolute', 1, PRECISION])
+        lines.append([attr_min, attr_min, 'absolute', 1, PRECISION])
+        attrlist.append(attr_max)
+        attrlist.append(attr_min)
+    elif get_sysfs_value(os.path.join(syspath, attr_min)) is not None:
+        lines.append([attr_min, attr_min, 'absolute', 1, PRECISION])
+        attrlist.append(attr_min)
+    if get_sysfs_value(os.path.join(syspath, attr_max_design)) is not None:
+        lines.insert(0, [attr_max_design, attr_max_design, 'absolute', 1, PRECISION])
+        lines.append([attr_min_design, attr_min_design, 'absolute', 1, PRECISION])
+        attrlist.append(attr_max_design)
+        attrlist.append(attr_min_design)
+    elif get_sysfs_value(os.path.join(syspath, attr_min_design)) is not None:
+        lines.append([attr_min_design, attr_min_design, 'absolute', 1, PRECISION])
+        attrlist.append(attr_min_design)
+    return {name: {'options': options, 'lines': lines}}, attrlist
+
+
+def get_charge_chart(syspath):
+    # Charge is measured in microamphours.  We track up to five
+    # attributes.
+    return get_generic_chart(syspath, 'charge', 'µAh', 'full', 'empty')
+
+
+def get_energy_chart(syspath):
+    # Energy is measured in microwatthours.  We track up to five
+    # attributes.
+    return get_generic_chart(syspath, 'energy', 'µWh', 'full', 'empty')
+
+
+def get_voltage_chart(syspath):
+    # Voltage is measured in microvolts. We track up to five attributes.
+    return get_generic_chart(syspath, 'voltage', 'µV', 'min', 'max')
+
+
+# This is a list of functions for generating charts.  Used below to save
+# a bit of code (and to make it a bit easier to add new charts).
+GET_CHART = {
+    'capacity': get_capacity_chart,
+    'charge': get_charge_chart,
+    'energy': get_energy_chart,
+    'voltage': get_voltage_chart
+}
+
+
+# This opens the specified file and returns the value in it or None if
+# the file doesn't exist.
+def get_sysfs_value(filepath):
+    try:
+        with open(filepath, 'r') as datasource:
+            return int(datasource.read())
+    except (OSError, IOError):
+        return None
+
+
+# Certain attributes (*_empty and *_empty_design) are implicitly zero
+# if not provided by the driver.  This function is used to fetch those
+# attributes.
+def get_sysfs_value_or_zero(filepath):
+    result = get_sysfs_value(filepath)
+    if result is None:
+        return 0
+    return result
```

Comment:
just `return get_sysfs_value(filepath) or 0`

Reply:
Or we can use `get_sysfs_value(filepath) or 0` where we need `get_sysfs_value_or_zero`